### PR TITLE
Use X-Forwarded-For header to define X-Real-IP header

### PR DESCRIPTION
**Pull request by zeph**
### What does this PR do?

It's implementing a mechanism similar to nginx to determine the x-real-ip by excluding proxy IPs

### Motivation

I stumbled on this issue while reviewing my freshly installed traefik v2, 
noticing that the x-real-ip header was not properly populated... #3097 
(I referred the feature request opened by @basro 2 years ago)

### More

I added a couple of test, the 1st to validate the feature
...the 2nd to avoid to break old behaviour

### Additional Notes

Also the logging shall be policed. Right not the `ClientAddr` is 
taken from the `X-Forwarded-For` just because it is present
...the proposed `X-Real-IP` would be a valid substitute

### CODE REVIEW

**Comment by zeph** [comment]
p.s. it was already failing out of the master I branched off... happy to rebase as soon as u give me a heads up

**Comment by zeph** [comment]
fails on origin/v2.4
```golang
FAIL: grpc_test.go:365: GRPCSuite.TestGRPCBufferWithFlushInterval

grpc_test.go:423:
    c.Assert(err, check.IsNil)
... value *fmt.wrapError = &fmt.wrapError{msg:"try operation failed: failed to receive stream data", err:(*errors.errorString)(0xc000e6e980)} ("try operation failed: failed to receive stream data")
```

**Comment by zeph** [comment]
failure on `master`

```golang
FAIL: docker_test.go:184: DockerSuite.TestDockerContainersWithLabels

docker_test.go:221:
    c.Assert(err, checker.IsNil)
... value *fmt.wrapError = &fmt.wrapError{msg:"try operation failed: got status code 404, wanted 200", err:(*errors.errorString)(0xc000951140)} ("try operation failed: got status code 404, wanted 200")
```

**Comment by ldez** [comment]
I guess that you are trying to run the integration tests locally.
To do that:
- first: `make pull-images` (only once)
- then `CI=1 make test-integration`

**Comment by zeph** [comment]
@ldez I'm indeed... since I wanna see my PR passing xD
...trying with an empty image set, lets see... `CI=1` this time

**Comment by zeph** [comment]
ok, all green even on my workstation... 
thanks for the heads-up @ldez & for kicking the CI process again :)

**Comment by jbdoumenjou** [comment]
Hello,
 
Thanks for your contribution!

This PR is now labeled `needs-design-review`, 
as you can see in our [contribution guide](https://github.com/traefik/contributors-guide/blob/master/review_guidelines.md), this means that we have to take a closer look at the scope of impact for that PR to see, how it would interact with the other parts of Traefik.

We will come back to you soon when we did the first design review iteration.

**Comment by zeph** [comment]
@jbdoumenjou I stepped on this issue this morning https://github.com/traefik/traefik/issues/7884
...eventually this `X-Real-IP` properly populated could be the new reference for the Middleware `IpWhitelist`
and make https://doc.traefik.io/traefik/middlewares/ipwhitelist/#ipstrategyexcludedips obsolete

**Comment by zeph** [comment]
@jbdoumenjou I'll wait ur heads-up before doing a rebase, seems that the merge is not possible automatically anymore

**Comment by made2591** [comment]
Any update on this?

**Comment by jbdoumenjou** [comment]
Hello @zeph,

We had a first look and want to clarify the purpose of the PR.
It seems that the addressed case here is when Traefik trusts the previous proxy that provides a `X-Forwarded-For` header but no `X-Real-Ip` header.
Currently, when the `X-Real-Ip` header is not set, Traefik creates it using the remote IP address.

This PR uses the `X-Forwarded-For` as a source to populate the `X-Real-Ip`, as it should contain the chain of previous client IPs.
Looking at the implementation, it uses the `forwardedHeader.trustedIPs` list to ignore all trusted IPs in the `X-Forwarded-For` and apply the first "non trusted" IP as `X-Real-Ip` header value.
This `trustedIPs` option is currently used to define the IPs (the request remote address) from which the forwarded headers will be trusted, so not deleted and used.

Why using this option, which is not expected, to filter out the IPs from the `X-Forwarded-For` to find the `X-Real-Ip` value? It does not seem to be appropriate.

Moreover, if the previous proxy is trusted, the `X-Forwarded-For` header should be trusted too and the `X-Real-Ip` value should be the first IP of the `X-Forwarded-For`.

WDYT?

**Comment by zhiguangwang** [comment]
@jbdoumenjou This PR is useful when you put traefik behind an AWS ALB (Application Load Balancer).

In this case, the request flows like this:

```
Client => ALB => Traefik => Application
```

Because [ALB appends client IP to X-Forwarded-For](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html#x-forwarded-for), when the request reaches the application, the HTTP headers becomes:

```
X-Forwarded-For: <client_ip>, <alb_ip>
X-Real-Ip: <alb_ip>
```

Further more, the client can forge `X-Forwarded-For`, so the HTTP headers becomes:

```
X-Forwarded-For: <fake_client_ip> <client_ip>, <alb_ip>
X-Real-Ip: <alb_ip>
```

But what we are really expecting is:

```
X-Forwarded-For: <client_ip>, <alb_ip>
X-Real-Ip: <client_ip>
```

**Comment by jbdoumenjou** [comment]
Hello,

Thanks to this PR, we realized that we have several concerns about the X-Real-Ip expected behavior.
Moreover, this implementation will not be backward compatible as it will change the current behavior.

So we will close this PR and continue the discussion in the original issue #3097

**Comment by zeph** [comment]
@jbdoumenjou can you make some examples of configurations that are going to break if introducing this "new behaviour", please?

**Comment by zeph** [comment]
@zhiguangwang with my patch the `<alb_ip>` is going to be removed from the list if it has been passed in the "trusted proxies" list
-> it is not going to be presented as "<real ip>" as in your example

**Comment by zeph** [comment]
@zhiguangwang to clarify... ok, in my infrastructure I have 2 reverse proxies in front of my traefik ... I remove from the game the 1st set with the `trustIPs` as main cluster configuration... and the 2nd instead is addressed when doing real IP filtering with this middleware 

```
$ cat templates/ipwhitelist.yml 
apiVersion: traefik.containo.us/v1alpha1
kind: Middleware
metadata:
  name: geo
spec:
  ipWhiteList:
    sourceRange:
 [ ... trusted 3rd parties ... ]

    ipStrategy:
      # https://doc.traefik.io/traefik/middlewares/ipwhitelist/#ipstrategyexcludedips
      excludedIPs:
        - 172.31.7.2/32 # 217.115.64.70,.71,.72 SNAT
        # https://www.cloudflare.com/ips-v4
        - 103.21.244.0/22
        - 103.22.200.0/22
        - 103.31.4.0/22
        - 104.16.0.0/12
        - 108.162.192.0/18
        - 131.0.72.0/22
        - 141.101.64.0/18
        - 162.158.0.0/15
        - 172.64.0.0/13
        - 173.245.48.0/20
        - 188.114.96.0/20
        - 190.93.240.0/20
        - 197.234.240.0/22
        - 198.41.128.0/17
        # https://www.cloudflare.com/ips-v6
        - 2400:cb00::/32
        - 2606:4700::/32
        - 2803:f800::/32
        - 2405:b500::/32
        - 2405:8100::/32
        - 2c0f:f248::/32
        - 2a06:98c0::/29
```

in your case... you need to exclude the ip ranges declared by AWS as the subnets for your `<alb ip>`

**Comment by zeph** [comment]
@zhiguangwang I pass the trustedIPs with this... in this case the IP of the 1st reverse proxy
```
    # https://community.traefik.io/t/how-to-properly-use-x-forwarded-for-in-v2/3834
    - "--entrypoints.websecure.forwardedHeaders.trustedIPs=x.x.x.x/32"
    - "--entrypoints.web.forwardedHeaders.trustedIPs=x.x.x.x/32"

```

**Comment by jbdoumenjou** [comment]
@Zeph,

The [trustedIPs](https://doc.traefik.io/traefik/routing/entrypoints/#forwarded-headers) is a way to trust the forwarded headers sent by a Client by checking its address. Therefore, It could be defined to trust only specific proxy/servers.

Using the trustedIPs option to set the X-Real-IP is a workaround and could impact the other forwarded headers as Traefik could trust unwanted servers. For example, as soon as one trusted proxy reaches Traefik and does not provide the X-Real-IP header, the X-Real-IP value would be then different from the proxy IP. We understand that this is the behavior you were expecting, thus it is a breaking change as the X-Real-IP value will change without modifying the existing configuration.

As said in the original issue (#3097), the current implementation is naive/simple but there is no real specification to oppose to it that would define the expected behavior.

In other words, we want to keep the contract for X-Real-IP header identical as for forwarded headers, the chain of trust is enforced by trusting or not the client which is reaching Traefik, and as for now we think that more specific/custom behaviors/capabilities should be handled with plugins and not by Traefik itself.
